nmpolicy/internal/parser/parser.go

```diff
@@ -20,64 +20,151 @@
 package parser
 
 import (
-	"reflect"
+	"fmt"
 
 	"github.com/nmstate/nmpolicy/nmpolicy/internal/ast"
+
 	"github.com/nmstate/nmpolicy/nmpolicy/internal/lexer"
 )
 
-type Parser struct{}
+type Parser struct {
+	tokens          []lexer.Token
+	currentTokenIdx int
+	lastNode        *ast.Node
+}
+
+func New() *Parser {
+	return &Parser{}
+}
+
+func (p *Parser) Parse(tokens []lexer.Token) (ast.Node, error) {
+	p.tokens = tokens
+	node, err := p.parse()
+	if err != nil {
+		return ast.Node{}, err
+	}
+	return node, nil
+}
+
+func (p *Parser) parse() (ast.Node, error) {
+	for {
+		if p.currentToken() == nil {
+			return ast.Node{}, nil
+		} else if p.currentToken().Type == lexer.EOF {
+			break
+		} else if p.currentToken().Type == lexer.STRING {
+			str := p.parseString()
+			p.lastNode = &str
+		} else if p.currentToken().Type == lexer.IDENTITY {
+			identity := p.parseIdentity()
+			p.lastNode = &identity
+			path, err := p.parsePath()
+			if err != nil {
+				return ast.Node{}, err
+			}
+			p.lastNode = path
+		} else if p.currentToken().Type == lexer.EQFILTER {
+			eqfilter, err := p.parseEqFilter()
+			if err != nil {
+				return ast.Node{}, err
+			}
+			p.lastNode = eqfilter
+		} else {
+			return ast.Node{}, &InvalidExpressionError{fmt.Sprintf("unexpected token `%+v`", p.currentToken().Literal)}
+		}
+		p.nextToken()
+	}
+	return *p.lastNode, nil
+}
+
+func (p *Parser) nextToken() {
```

Comment:
I would expect `nextToken` to return a token. Similar to `prevToken`.

Reply:
Aligning `prevToken` to `nextToken` returning values is not used so I remove it.